src/backend/distributed/connection/remote_commands.c

```diff
@@ -1042,6 +1042,12 @@ SendCancelationRequest(MultiConnection *connection)
 {
 	char errorBuffer[ERROR_BUFFER_SIZE] = { 0 };
 
+	if (!PQisBusy(connection->pgConn))
```

Comment:
I don't quite understand the relationship between the checks in ShutdownConnection vs PQisBusy. Do we still need the other checks?

Reply:
that is a good question. I haven't really looked at the checks in shutdown to understand what they are testing. I settled on `PQisBusy` to determine to send the cancel due to the comment on the state `PQisBusy` checks for:

https://github.com/postgres/postgres/blob/401aad67045b2d467571b54abe229fdd115a228c/src/interfaces/libpq/libpq-int.h#L219

There are other states as well, like copy, however copies can't be cancelled anyway.

The check I added in `ShutdownConnection` is to not change the shutdown connection if there was nothing to cancel. Although I believe the waiting on the result ignoring all errors uses `PQisBusy` as well.

I will have a look at the last remaining checks, and minimise the checks if possible.

Reply:
It seems the main difference between PQisBusy and the existing checks does a parseInput, which might be useful. It would be preferable to keep them in one place.

I think an additional check that might be useful is `connection->remoteTransaction.transactionState != REMOTE_TRANS_SENT_BEGIN` to avoid cancelling BEGIN.